zebra: Add the support of the Source Addr param of the SRv6 Encapsulation

cscarpitta opened the pull request:
In SRv6, the customer packet is encapsulated with an outer IPv6 header and optional SRH. The DA of the outer IPv6 header is set to the first SID from the SID List. The Source address of the IPv6 header is set to an IPv6 address configured by the operator (i.e. the source of the SRv6 tunnel). The configuration of the SRv6 encapsulation source address is missing in FRR. 

In this PR, we extend Zebra to support the configuration of the SRv6 encapsulation source address. We added the required vtysh command, topotest, and documentation changes. 

The SRv6 encapsulation source address can be configured in FRR using the CLI command shown below:

```
router(config)# segment-routing
router(config-sr)# srv6
router(config-srv6)# encapsulation
router(config-srv6-encap)# source-address fc00:0:1::1
```

**Note for Linux:** 

The Linux dataplane implementation already supports the configuration of the SRv6 encapsulation source address. The following iproute2 command can be used to configure the source address. The iproute2 command uses a Generic Netlink message. 

```
# ip sr tunsrc set  fc00:1::1
```

If the user didn’t set the SRv6 Encapsulation source address, the kernel uses the IPv6 address of the outgoing interface. However, this causes a degradation in the packet forwarding performance for the SRv6 Encapsulation behavior in the Kernel. The degradation has been studied in the SRv6 Performance benchmarking published in https://ieeexplore.ieee.org/document/9311210


**Note for SONiC:**

The SRv6 Encapsulation Source address is one of the parameters that must be configured in two of the SONiC AppDB tables (SRV6_MY_SID_TABLE, ROUTE_TABLE) in the case of SRv6.
 
The absence of the source address configuration in SONiC AppDB causes SONiC to crash. 

Signed-off-by: Carmine Scarpitta carmine.scarpitta@uniroma2.it

ton31337 commented on zebra/ge_netlink.c:
```diff
@@ -0,0 +1,237 @@
+			 strlen(family) + 1))
+		return 0;
+
+	if (!strcmp(family, "SEG6"))
```
nit: strmatch (easier to read)

cscarpitta replied:
Fixed, thanks.

ton31337 commented on zebra/ge_netlink.c:
```diff
@@ -0,0 +1,237 @@
+	struct nlsock *nl;
+	const struct in6_addr *tunsrc_addr;
+
+	struct {
```
Maybe move this struct to a header file, because it's reused already in a couple of places?

cscarpitta replied:
Moved to ge_netlink.h, thanks.

ton31337 commented on zebra/zebra_srv6.c:
```diff
@@ -369,6 +369,20 @@ int release_daemon_srv6_locator_chunks(struct zserv *client)
+{
+	struct zebra_srv6 *srv6 = zebra_srv6_get_default();
+
+	memcpy(&srv6->encap_src_addr, encap_src_addr, sizeof(struct in6_addr));
```
Should we check against NULL for the source?

ton31337 commented on zebra/zebra_srv6_vty.c:
```diff
@@ -76,6 +76,71 @@ static struct cmd_node srv6_loc_node = {
+	.prompt = "%s(config-srv6-encap)# "
+};
+
+DEFUN (show_srv6,
```
DEFPY

cscarpitta replied:
Fixed, thanks.

ton31337 commented on zebra/zebra_srv6_vty.c:
```diff
@@ -76,6 +76,71 @@ static struct cmd_node srv6_loc_node = {
+
+		id = 1;
+		for (ALL_LIST_ELEMENTS_RO(srv6->locators, node, locator)) {
+			prefix2str(&locator->prefix, str, sizeof(str));
```
`%pFX`

cscarpitta replied:
Fixed, thanks.

ton31337 commented on zebra/zebra_srv6_vty.c:
```diff
@@ -345,6 +410,37 @@ DEFPY (locator_prefix,
+
+DEFPY (no_srv6_src_addr,
+       no_srv6_src_addr_cmd,
+       "no source-address",
```
We should allow optionally specifying `X:X::X:X` source address.

cscarpitta replied:
Changed the command to `no source-address [X:X::X:X$encap_src_addr]`. Thanks.

ton31337 reviewed (commented):
Some nits.

cscarpitta commented on doc/user/zebra.rst:
```diff
@@ -787,6 +849,14 @@ and this section also helps that case.
+
+   Configure parameters for SRv6 encapsulation.
+
+.. clicmd:: [no] source-address X:X::X:X
```
Dropped, thanks.

cscarpitta commented on zebra/ge_netlink.c:
```diff
@@ -0,0 +1,237 @@
+	if (h->nlmsg_type != GENL_ID_CTRL) {
+		zlog_err(
+			"%s: Not a controller message, nlmsg_len=%d nlmsg_type=0x%x",
+			__func__, h->nlmsg_len, h->nlmsg_type);
```
Fixed, thanks.

cscarpitta commented on zebra/kernel_netlink.c:
```diff
@@ -1257,6 +1278,32 @@ int netlink_talk(int (*filter)(struct nlmsghdr *, ns_id_t, int startup),
+	/* Capture info in intermediate info struct */
+	dp_info.ns_id = zns->ns_id;
+
+#if defined(HAVE_NETLINK)
```
Fixed, thanks.

cscarpitta replied:
I realized that the whole `kernel_netlink.c` file is contained in a `#ifdef HAVE_NETLINK` block.
Therefore, checking for `HAVE_NETLINK` in the `ge_netlink_talk()` function is redundant. I removed this check.

cscarpitta commented on zebra/kernel_netlink.c:
```diff
@@ -1853,6 +1915,15 @@ void kernel_init(struct zebra_ns *zns)
+
+	if (ret < 0)
+		zlog_notice(
+			"Registration for extended generic netlink cmd ACK failed : %d %s",
```
Right! Replaced `zlog_notice()` -> `zlog_err()`, thanks. :)

cscarpitta commented on zebra/zebra_srv6_vty.c:
```diff
@@ -356,6 +452,12 @@ static int zebra_sr_config(struct vty *vty)
 	if (zebra_srv6_is_enable()) {
 		vty_out(vty, "segment-routing\n");
 		vty_out(vty, " srv6\n");
+		if (memcmp(&srv6->encap_src_addr, &in6addr_any,
```
Fixed, thanks.

cscarpitta commented on zebra/zebra_dplane.c:
```diff
@@ -5027,6 +5053,59 @@ dplane_gre_set(struct interface *ifp, struct interface *ifp_link,
+					  1, memory_order_relaxed);
+		if (ctx)
+			dplane_ctx_free(&ctx);
+		result = ZEBRA_DPLANE_REQUEST_FAILURE;
```
Right. By default, result is `ZEBRA_DPLANE_REQUEST_FAILURE`. Therefore, this line is useless.
Removed, thanks.

cscarpitta commented on zebra/kernel_netlink.c:
```diff
@@ -1257,6 +1278,32 @@ int netlink_talk(int (*filter)(struct nlmsghdr *, ns_id_t, int startup),
+int ge_netlink_talk(int (*filter)(struct nlmsghdr *, ns_id_t, int startup),
+		    struct nlmsghdr *n, struct zebra_ns *zns, bool startup)
+{
+	struct zebra_dplane_info dp_info;
```
It is not an issue because the functions defined in `kernel_netlink.c` exist only if `HAVE_NETLINK`, as stated in https://github.com/FRRouting/frr/pull/12261#discussion_r1015970668.

If no `HAVE_NETLINK`, `ge_netlink_talk` will not exist and cannot use uninitialized variables :)

cscarpitta commented:
@ton31337 Thank you for the review, Donatas.
I addressed all your comments, rebased and force-pushed.

cscarpitta commented:
Rebased to resolve merge conflicts.

donaldsharp commented on zebra/kernel_netlink.c:
```diff
@@ -386,6 +386,27 @@ static int netlink_socket(struct nlsock *nl, unsigned long groups,
 }
 
+/* Make socket for Linux generic netlink interface. */
+static int ge_netlink_socket(struct nlsock *nl, ns_id_t ns_id)
```
why are we not abstracting and using netlink_socket and why are we not using bind?  I'm not a fan of cut-n-paste coding.  If we need this then let's properly abstract it

cscarpitta replied:
@donaldsharp Many thanks for your review!

I made the requested change.

1. I abstracted `netlink_socket()`. Now `netlink_socket()` takes an additional argument `nl_family` that specifies the Netlink family of the socket to create. This allows the developer to create any kind of Netlink socket (not only `NETLINK_ROUTE` sockets).
2. I added a new function `rt_netlink_socket()` that can be used to create a `NETLINK_ROUTE` socket. It calls `netlink_socket()` with `nl_family=NETLINK_ROUTE`.
3. I added a new function `ge_netlink_socket()` that can be used to create a `NETLINK_GENERIC` socket. It calls `netlink_socket()` with `nl_family=NETLINK_GENERIC`.

Here is the diff: https://github.com/FRRouting/frr/compare/43430662781371e161d3084d452b5484900aef23..178b584374ae5691a8ce6252faf26550080760b8

cscarpitta commented:
@donaldsharp @ton31337 Thanks again for your review!

I made the changes requested by @donaldsharp in https://github.com/FRRouting/frr/pull/12261#discussion_r1024168146.

Here you can find the diff:
https://github.com/FRRouting/frr/compare/43430662781371e161d3084d452b5484900aef23..178b584374ae5691a8ce6252faf26550080760b8

ton31337 commented on doc/user/zebra.rst:
```diff
@@ -680,6 +680,68 @@ FRR's cli or frr.conf or zebra.conf. This section shows how
+      Locator:
+      Name                 ID      Prefix                   Status
+      -------------------- ------- ------------------------ -------
+      loc1                       1 fc00:0:1::/48            Up
```
just a nit: why is this aligned to the right, while others - left?

cscarpitta replied:
@ton31337 Because I missed a "-" in the `vty_out()` call :)

My intention is to align everything to the left, I think it looks nicer.
I updated both the VTY output and the documentation. Thanks.

ton31337 commented on zebra/ge_netlink.c:
```diff
@@ -0,0 +1,220 @@
+			return 0;
+		break;
+	default:
+		zlog_err("%s: unsupported command (%u)", __func__, cmd);
```
Please, drop `__func__`.

cscarpitta replied:
Dropped, thanks.

ton31337 commented on zebra/ge_netlink.c:
```diff
@@ -0,0 +1,220 @@
+	default:
+		/* Invalid op */
+		zlog_err(
+			"%s: context received for kernel sr tunsrc update with incorrect OP code (%u)",
```
Ditto regarding func.

cscarpitta replied:
Dropped, thanks.

ton31337 commented on zebra/zebra_srv6_vty.c:
```diff
@@ -404,6 +468,38 @@ DEFPY (locator_behavior,
+       "no source-address [X:X::X:X$encap_src_addr]",
+       NO_STR
+       "Segment Routing SRv6 source address\n"
+	   "Specify source address for SRv6 encapsulation\n")
```
Seems wrong indentation.

cscarpitta replied:
Fixed, thanks.

cscarpitta commented:
@ton31337 Thanks for your review. I addressed all your suggestions and force-updated the branch.

Here are the changes:
* Dropped `__func__` from the `zlog_err()` calls.
* Fixed an indentation issue in a `DEFPY`.
* Fixed a bad alignment in the VTY output.

Diff: https://github.com/FRRouting/frr/compare/c33ac26652f63550cca34f6c0ed151c8fb291830..4fe5fee6604037717cff1da83f7559912e5e90bb

cscarpitta commented:
@ton31337 I addressed all the comments, could you please take a look again?

cscarpitta commented:
Hi @donaldsharp  @ton31337 @eqvinox 

Following our discussion today, this PR extends FRR to support the Source address configuration for the SRv6 Encapsulation over the linux kernel.

The Linux kernel supports the configuration of the source address using the following ip route command.

`ip sr tunsrc set  fc00:1::1`

We reuse the same Netlink message for configuring the source address using FRR.

cscarpitta commented:
ci:rerun

cscarpitta commented:
@donaldsharp @ton31337 Many thanks for the review. All comments have been addressed.

donaldsharp commented on zebra/kernel_netlink.c:
```diff
@@ -382,6 +385,15 @@ static int netlink_socket(struct nlsock *nl, unsigned long groups,
 }
 
+/* Make socket for Linux route netlink interface. */
+static int rt_netlink_socket(struct nlsock *nl, unsigned long groups,
```
I would just remove this wrapper function and just add the NETLINK_ROUTE to the netlink_socket call.  Too much abstraction for very little gain

donaldsharp commented on zebra/kernel_netlink.c:
```diff
@@ -316,7 +316,7 @@ static int netlink_socket(struct nlsock *nl, unsigned long groups,
 	int namelen;
 
-	if (nl_family != NETLINK_ROUTE)
+	if (nl_family != NETLINK_ROUTE && nl_family != NETLINK_GENERIC)
```
Do we really need this test?  This is a developer problem that they are passing in a wrong nl_family

cscarpitta replied:
I removed this test.

donaldsharp commented on zebra/kernel_netlink.c:
```diff
@@ -394,6 +394,15 @@ static int rt_netlink_socket(struct nlsock *nl, unsigned long groups,
 }
 
+/* Make socket for Linux generic netlink interface. */
+static int ge_netlink_socket(struct nlsock *nl, unsigned long groups,
```
again too much abstraction.  Just call the netlink_socket with the NETLINK_GENERIC

donaldsharp commented on zebra/zebra_ns.h:
```diff
@@ -49,6 +49,8 @@ struct zebra_ns {
 	struct nlsock netlink_dplane_in;
 	struct event *t_netlink;
+
+	struct nlsock ge_netlink_cmd; /* command channel for generic netlink */
```
let's just use the zebra_router_get_next_sequence.  It will stay consistent with everything else we do.  FRR uses one sequence number generator for all it's netlink sockets.

donaldsharp commented on zebra/ge_netlink.c:
```diff
@@ -0,0 +1,111 @@
+
+static int16_t seg6_genl_family = -1;
+
+static int genl_parse_getfamily(struct nlmsghdr *h, ns_id_t ns_id, int startup)
```
What is exactly seg6_genl_family?  It should be retrieved on startup like all the other startup information.  So FRR doesn't have a special case that it's gathered on first attempt to talk to the data plane.

donaldsharp replied:
Additionally if retrieval on startup *fails* this can be communicated via the dplane that configuring this will not work properly.  IE -> the main pthread of zebra should be sent the idea of whether or not the generic netlink messaging works if the socket fails to create ( totally possible on older kernels ).  This way when someone enters the encapsulation commands it can be noted this ain't going to work on this kernel.

cscarpitta replied:
I made the required changes.

`seg6_genl_family` is a numeric identifier for "SRv6" and is used to send "SRv6" commands to the kernel over the Generic Netlink socket.

On startup, zebra queries the kernel to find out if kernel supports configuring SRv6 encap parameters through Generic Netlink and retrieves the `seg6_genl_family` identifier. If it is not supported by the kernel, any future attempt to set SRv6 encap parameters will return an error.

cscarpitta commented:
@donaldsharp Many thanks for the review. All comments have been addressed.

donaldsharp commented on zebra/kernel_netlink.c:
```diff
@@ -1958,6 +1959,9 @@ void kernel_init(struct zebra_ns *zns)
 
 	rt_netlink_init();
+
+	if (zns->ge_netlink_cmd.sock >= 0)
```
I'd move this if to inside the call to ge_netlink_init().  kernel_init() shouldn't care about the status.  It just knows it needs to ask the question.

cscarpitta replied:
Moved to `ge_netlink_init()`.

donaldsharp commented:
a) There is no handler for when another process changes the value in the kernel.  My assumption is that this can be listened for?
```
sharpd@eva:~/iproute2/ip$ sudo ip sr tunsrc show
[sudo] password for sharpd: 
tunsrc addr fc00:0:1::1
sharpd@eva:~/iproute2/ip$ sudo vtysh -c "show segment-routing srv6 manager"
Parameters:
  Encapsulation:
    Source Address:
      Configured: ::
```
b) There is no distinction for what happens when a kernel value is different than the configured value.
```
sharpd@eva:/work/home/sharpd/frr1/zebra$ sudo ip sr tunsrc show
tunsrc addr fc00:4::3
sharpd@eva:/work/home/sharpd/frr1/zebra$ sudo vtysh -c "show segment-routing srv6 manager"
Parameters:
  Encapsulation:
    Source Address:
      Configured: fc00:4::3
sharpd@eva:/work/home/sharpd/frr1/zebra$ sudo ip sr tunsrc set fc00:0:2::44
sharpd@eva:/work/home/sharpd/frr1/zebra$ sudo ip sr tunsrc show
tunsrc addr fc00:0:2::44
sharpd@eva:/work/home/sharpd/frr1/zebra$ sudo vtysh -c "show segment-routing srv6 manager"
Parameters:
  Encapsulation:
    Source Address:
      Configured: fc00:4::3
```

cscarpitta commented:
@donaldsharp

I updated the PR to do the following:

- On startup, we retrieve the actual encap source address from the kernel and save this address in zebra config
- When user changes the encap source address in zebra config -> we push the new address to the kernel

As we discussed, actually we cannot listen for changes of the encap source address in the kernel, because the kernel does not send us any notification when the value changes. 

I added a comment in the code to mention that this is currently not supported:

>  At the time of writing this code, the kernel does not send us any asynchronous
>  notifications when someone changes the 'sr tunsrc' under us. As a result, we
>  are currently unable to detect when the source address changes and update the
>  SRv6 encapsulation source address configured in zebra.
>  
>  In the future, when the kernel supports async notifications, the implementation
>  can be improved by listening on the Generic Netlink socket and adding a handler
>  to process/parse incoming 'sr tunsrc' change messages and update the SRv6 zebra
>  configuration with the new encap source address.

I will submit another PR to extend this functionality and handle kernel notifications for source address changes, as soon as this is supported by the kernel.

cscarpitta commented:
Thanks for the review. All comments have been addressed.

riw777 reviewed (approved):
looks good ... will wait a few days to see if anyone else wants to review

cscarpitta commented:
ci:rerun unrelated failures, rerunning

cscarpitta commented:
> looks good ... will wait a few days to see if anyone else wants to review

@riw777 Many thanks for the review. I rebased on top of latest master to solve conflicts.
